src/common.cpp

```diff
@@ -1149,161 +1149,185 @@ maybe_t<size_t> read_unquoted_escape(const wchar_t *input, wcstring *result, boo
-                    if (0x10FFFF < max_val) max_val = static_cast<wchar_t>(0x10FFFF);
-                    break;
-                }
-                case L'x': {
-                    chars = 2;
-                    max_val = ASCII_MAX;
-                    break;
-                }
-                case L'X': {
-                    byte_literal = true;
-                    max_val = BYTE_MAX;
-                    break;
+                        // Don't exceed the largest Unicode code point - see #1107.
+                        if (0x10FFFF < max_val) max_val = static_cast<wchar_t>(0x10FFFF);
+                        break;
+                    }
+                    case L'x': {
+                        chars = 2;
+                        max_val = ASCII_MAX;
+                        break;
+                    }
+                    case L'X': {
+                        byte_literal = true;
+                        max_val = BYTE_MAX;
+                        break;
+                    }
+                    default: {
+                        base = 8;
+                        chars = 3;
+                        // Note that in_pos currently is just after the first post-backslash
+                        // character; we want to start our escape from there.
+                        assert(in_pos > 0);
+                        in_pos--;
+                        break;
+                    }
                 }
-                default: {
-                    base = 8;
-                    chars = 3;
-                    // Note that in_pos currently is just after the first post-backslash character;
-                    // we want to start our escape from there.
-                    assert(in_pos > 0);
-                    in_pos--;
-                    break;
+
+                for (size_t i = 0; i < chars; i++) {
+                    long d = convert_digit(input[in_pos], base);
+                    if (d < 0) {
+                        // If we have no digit, this is a tokenizer error.
+                        if (i == 0) errored = true;
+                        break;
+                    }
+
+                    res = (res * base) + d;
+                    in_pos++;
                 }
-            }
 
-            for (size_t i = 0; i < chars; i++) {
-                long d = convert_digit(input[in_pos], base);
-                if (d < 0) {
-                    // If we have no digit, this is a tokenizer error.
-                    if (i == 0) errored = true;
-                    break;
+                if (!errored && res <= max_val) {
+                    if (byte_literal) {
+                        // Multibyte encodings necessitate that we keep adjacent byte escapes.
+                        // - `\Xc3\Xb6` is "ö", but only together.
+                        // (this assumes a valid codepoint can't consist of multiple bytes
+                        // that are valid on their own, which is true for UTF-8)
```

Comment:
It's true for most multi-byte encodings but the opposite isn't (e.g. instead of `\Xc3\Xb6` this could just as well be `\X51\X51` each of which stands by itself as a character). This'll cause you to keep trying to concatenate adjacent single-byte (or multiple multi-byte) escapes to `byte_buff` at no point stopping to return them upstream... which as far as I can tell, is OK in the general case.

It does mean if an error is encountered parsing a subsequent adjacent escape after consuming one or more bytes that fully form one (or more) character(s), you'll error out from the first escape instead of just the actually broken one, with none of the escaped characters making into the valid payload buffer. I don't know if that matters.

Actually, after testing, it seems that's already the case and pedantically we already report an incorrect error. `fish -c 'echo \\X51\\X'` returns `fish: Invalid token '\X51\X' ...` instead of `fish: Invalid token '\X'`.

Reply:
>even if it's suboptimal because you might exceed the SSO for byte_buff

If we have been given a large string, we need to handle a large string.

Technically we could force a str2wcstring and clear the buffer after 4 bytes, which should be the limit on any sensible encoding (even UTF-32, the simplest unicode encoding), but I didn't really want to give this function that knowledge, and it seems like an optimization for a special case that very very likely is never going to happen.

If you want to test, see how a file with thousands of `\Xab` in a row fares. My expectation is that it is slower than just using the characters immediately, but not terrible.

Reply:
I guess we could use MB_CUR_MAX.

Reply:
Okay, in any realistic case this simply does *not matter*.

My test case is a file created with

```fish
echo echo (string repeat -n 2000000 "\\X82") >echo.fish
```

This will look like

```fish
echo \X82\X82\X82\X82...
```

The timing here is ~500ms without the decoding (i.e. current master), ~720ms with this PR and ~740ms if I add a `byte_buff.size() <= MB_CUR_MAX`.

This sounds like a lot, but it's also a *completely ridiculous file*, and current master is faster *but also wrong*, and flushing the decode early is slower. This isn't something where you can find a short and realistic script that hits a weird edge case and is super slow, this is something where you need to explicitly construct something ridiculous and it's still only slower by a constant factor - it's ~40% *if all you do is reading `\X`*.

So I do not believe this is something where the performance is different in any important way, and we should simply prefer correctness and simplicity.

Reply:
Yeah, I wasn't suggesting we should do anything about it once I observed that there were no functional regressions. I think it's fine the way it is unless there is some edge case I am missing that _doesn't_ involve erroring out where `read_unquoted_escape()` needs to return exactly one codepoint and no more. Not having to do anything that requires introspection of the current codepage is definitely a win.

Reply:
>read_unquoted_escape() needs to return exactly one codepoint and no more

Not that it matters, but: It can return multiple, and it will, here - or we would have to know codepoint boundaries, which requires knowing the encoding.

Reply:
That's what I was trying to say, but I guess didn't do a good job of it :)